packages/flutter/lib/src/painting/decoration.dart

```diff
@@ -44,27 +44,72 @@ abstract class Decoration extends Diagnosticable {
+  ///
+  /// For example, when a [BoxDecoration] has [BoxShape.circle], the padding
+  /// does not take into account that the circle is drawn in the center of the
+  /// box regardless of the ratio of the box; it does not provide the extra
+  /// padding that is implied by changing the ratio.
+  EdgeInsets get padding => EdgeInsets.zero;
 
   /// Whether this decoration is complex enough to benefit from caching its painting.
   bool get isComplex => false;
 
-  /// Linearly interpolates from [a] to [this].
-  Decoration lerpFrom(Decoration a, double t) => this;
+  /// Linearly interpolates from `a` to [this].
+  ///
+  /// When implementing this method in subclasses, return null if this class
+  /// cannot interpolate from `a`. In that case, [lerp] will try `a`'s [lerpTo]
+  /// method instead.
+  ///
+  /// Supporting interpolating from null is recommended as the [Decoration.lerp]
+  /// method uses this as a fallback when two classes can't interpolate between
+  /// each other.
+  ///
+  /// Instead of calling this directly, use [Decoration.lerp].
+  @protected
+  Decoration lerpFrom(Decoration a, double t) => null;
 
-  /// Linearly interpolates from [this] to [b].
-  Decoration lerpTo(Decoration b, double t) => b;
+  /// Linearly interpolates from [this] to `b`.
+  ///
+  /// This is called if `b`'s [lerpTo] did not know how to handle this class.
+  ///
+  /// When implementing this method in subclasses, return null if this class
+  /// cannot interpolate from `b`. In that case, [lerp] will apply a default
+  /// behavior instead.
+  ///
+  /// Supporting interpolating to null is recommended as the [Decoration.lerp]
+  /// method uses this as a fallback when two classes can't interpolate between
+  /// each other.
+  ///
+  /// Instead of calling this directly, use [Decoration.lerp].
+  @protected
+  Decoration lerpTo(Decoration b, double t) => null;
 
-  /// Linearly interpolates from [begin] to [end].
+  /// Linearly interpolates from `begin` to `end`.
   ///
-  /// This defers to [end]'s [lerpTo] function if [end] is not null,
-  /// otherwise it uses [begin]'s [lerpFrom] function.
+  /// This defers to `end`'s [lerpTo] function if `end` is not null. If `end` is
+  /// null or if its [lerpTo] returns null, it uses `begin`'s [lerpFrom]
+  /// function instead. If both return null, it attempts to lerp from `begin` to
+  /// null before `t=0.5`, then null to `end` after `t=0.5`.
   static Decoration lerp(Decoration begin, Decoration end, double t) {
+    Decoration result;
     if (end != null)
-      return end.lerpFrom(begin, t);
-    if (begin != null)
-      return begin.lerpTo(end, t);
-    return null;
+      result = end.lerpFrom(begin, t);
+    if (result == null && begin != null)
+      result = begin.lerpTo(end, t);
+    if (result == null && begin != null && end != null) {
+      if (t < 0.5) {
+        result = begin.lerpTo(null, t * 2.0);
+      } else {
+        result = end.lerpFrom(null, (t - 0.5) * 2.0);
```

Comment:
Why `(t - 0.5) * 2.0`?  Is it intended that with `t==0.5`, this will be zero?

Reply:
Yes, unless I'm thinking about it wrong.

It's intended to go map 0..0.5 to 0..1 for the lerpTo null, and 0.5..1.0 to 0.0..1.0 for the lerpFrom null. That way the animation from 0 to 1 overall goes from begin to null to end.